### DISCUSSION TITLE
Error while receiving logs

### DISCUSSION THREAD

**Discussion by pavellety** [cirruslabs/cirrus-cli#643] [Q&A] [answered]
Hi,

First of all thank you very much for woking on tart/cirrus-cli. This is wonderful project.

Anyway I have one question regarding the error we started to see quite often on our workers: `17:15:58  Error while receiving logs: rpc error: code = Canceled desc = context canceled.` 

I have tracked the implementation is here:
```
logEntry, err := stream.Recv()
                if err == io.EOF {
                        break
                }
                if err != nil {    
                        streamLogger.Warnf("Error while receiving logs: %v", err)
                        return err 
                }
```
Usually when this error occurs the job will stuck and fail after the configured timeout is reached (resources are blocked). 

Does it mean the client is dead? 
Should the task immediately fail when this error occurs? 
Would it make sense to implement retry mechanism (like fail after certain number of tries)?

We saw some network glitches in our monitoring on that workers so we suspect it can be related to this.

Thank you very much for reply.
Pavel

**Comment by edigaryev** [comment]
Hello Pavel!

Do I understand correctly that you use `cirrus run` to run your task on worker machines (and not [Persistent Workers](https://cirrus-ci.org/guide/persistent-workers/))?

The error you're seeing should not normally happen and likely means that the agent running the task has terminated/hung prematurely.

Which environment this happens on (Docker container, Tart VM, etc.)?

**Reply by pavellety** [reply]
Thank you for a reply.

You are correct we do not use Persistent Workers. Currently (Unfortunately) we do have a Jenkins layer which generates `.cirrus.yaml` and `cirrus run` command the job is then delegated to one of our physical M1/M2 machine  where it's executed using Tart isolation. 

Regarding environment:
```
tart: 1.5.0
cirrus-cli: 0.101.1-63e767b
Worker OS: 
        ProductName:		macOS 
        ProductVersion:		13.0 
        BuildVersion:		22A8381
        22.1.0 Darwin Kernel Version 22.1.0: Sun Oct  9 20:19:47 PDT 2022; root:xnu-8792.41.9~3/RELEASE_ARM64_T8112 arm64
```
Let me know if further information are needed. 

Thanks

**Reply by pavellety** [reply]
Any update here? Thanks

**Reply by edigaryev** [reply]
Hello Pavel,

Sorry for late response.

If I understood you right, your primary issue is that the `cirrus run` terminates because of the task's timeout, but not because of the networking error you're seeing, which could happen much earlier?

We could add a logic to fail the task if the agent misses more than 2 heartbeats (which happen every minute), but that won't fix the networking-related root cause you're experiencing.

**Reply by pavellety** [reply]
Hi,

thanks for the reply,

If there is not any way how tart could recover from this type of issue I would definitely prefer task to fail as you suggested.

Thanks,
Pavel

**Reply by edigaryev** [reply, marked as answer]
I've created https://github.com/cirruslabs/cirrus-cli/pull/652 to address this.

**Reply by edigaryev** [reply]
Please check out the new [`1.104.0` release](https://github.com/cirruslabs/cirrus-cli/releases/tag/v0.104.0) that include theses changes.

**Reply by edigaryev** [reply]
Hello @pavellety, are you still experiencing network glitches?

We've tweaked things a bit in the latest [`1.106.0` release](https://github.com/cirruslabs/cirrus-cli/releases/tag/v0.106.0), and now you'll need to explicitly enable heartbeat monitoring with `--heartbeat-timeout` flag.

The old behavior can be replicated by passing a `--heartbeat-timeout=2m` flag to `tart run`.

**Reply by pavellety** [reply]
Hello @edigaryev, let me try to bump to the new version and try. Thanks
